tempo_embeddings/text/abstractcorpus.py

```diff
@@ -0,0 +1,296 @@
+                see Corpus.tfidf_vectorizer() and Corpus.topic_words().
+                If None, uses most words with largest document frequency.
+            exclude_word: The word to exclude from the label,
+                e.g. the search term used for composing this corpus
+            n: concatenate the top n words in the corpus as the label.
+            stopwords: if given, exclude these words
+        """
+        # TODO: merge exclude_word and stopwords?
+
+        if exclude_words is None:
+            exclude_words = set()
+        exclude_words = [word.casefold() for word in exclude_words]
+
+        if vectorizer:
+            words = self._tf_idf_words(vectorizer, n=n * 2)
+        else:
+            words = self._document_frequencies(n=None)
+
+        top_words: list[str] = [
+            word
+            for word in words
+            if word.strip() and word.casefold() not in exclude_words
+        ]
+
+        self._label = "; ".join(top_words[:n])
+
+        return self._label
+
+    def _tf_idf_words(self, vectorizer: TfidfVectorizer, n: int = None) -> list[str]:
+        """The most words in the corpus with maximum <tf-idf> x <distance to centroid>.
+
+        Each word is scored by multiplying its tf-idf score for each passage
+        with the passage's embedding distance to the centroid.
+
+        Args:
+            vectorizer: The vectorizer to use for tf-idf scoring;
+                typically generated by calling Corpus.tfidf_vectorizer()
+                on a super-corpus (ie. a superset of this collection).
+            n: The number of words to return. Defaults to None, returning all words
+
+        Returns:
+            the n highest scoring words in a sorted list
+        """
+        tf_idfs: csr_matrix = self.tf_idf(vectorizer)
+
+        assert all(
+            passage.highlighting for passage in self.passages
+        ), "Passages must have highlightings"
+
+        assert tf_idfs.shape == (
+            len(self.passages),
+            len(AbstractCorpus.get_vocabulary(vectorizer)),
+        ), f"tf_idfs shape ({tf_idfs.shape}) does not match expected shape."
+
+        ### Weigh in vector distances
+        distances: ArrayLike = self.distances(normalize=True)
+        weights = np.ones(distances.shape[0]) - distances
+
+        assert weights.argmin() == distances.argmax()
+        assert weights.argmax() == distances.argmin()
+        assert (
+            weights.shape[0] == tf_idfs.shape[0]
+        ), f"distances shape ({weights.shape}) does not match expected shape."
+
+        weighted_tf_idfs = np.average(tf_idfs.toarray(), weights=weights, axis=0)
+        assert weighted_tf_idfs.shape[0] == len(
+            AbstractCorpus.get_vocabulary(vectorizer)
+        )
+
+        # pylint: disable=invalid-unary-operand-type
+        top_indices = np.argsort(-weighted_tf_idfs)[:n]
+
+        return [AbstractCorpus.get_vocabulary(vectorizer)[i] for i in top_indices]
```

Comment:
It is expensive, but I have resolved this by caching the outcome (`@lru_cache`). That way, it can
1. be called from different places and still reuse previously computed results and
2. have the caching in place for different vectorizers if needed

Reply:
Correct. I only brought this up because currently "get_vocabulary" is only executed in `_tf_idf_words`, but I should have also foreseen that this function might be called elsewhere.